website/docs/language/expressions/custom-conditions.mdx

```diff
@@ -0,0 +1,353 @@
+
+```hcl
+
+resource "aws_instance" "example" {
+  instance_type = "t2.micro"
+  ami           = "ami-abc123"
+
+  lifecycle {
+    # The AMI ID must refer to an AMI that contains an operating system
+    # for the `x86_64` architecture.
+    precondition {
+      condition     = data.aws_ami.example.architecture == "x86_64"
+      error_message = "The selected AMI must be for the x86_64 architecture."
+    }
+
+    # The EC2 instance must be allocated a private DNS hostname.
+    postcondition {
+      condition     = self.private_dns != ""
+      error_message = "EC2 instance must be in a VPC that has private DNS hostnames enabled."
+    }
+  }
+}
+
+data "aws_ebs_volume" "example" {
+  # Use data resources that refer to other resources to
+  # load extra data that isn't directly exported by a resource.
+  #
+  # Read the details about the root storage volume for the EC2 instance
+  # declared by aws_instance.example, using the exported ID.
+
+  filter {
+    name = "volume-id"
+    values = [aws_instance.example.root_block_device.volume_id]
+  }
+}
+
+output "api_base_url" {
+  value = "https://${aws_instance.example.private_dns}:8433/"
+
+  # The EC2 instance will have an encrypted root volume.
+  precondition {
+    condition     = data.aws_ebs_volume.example.encrypted
+    error_message = "The server's root volume is not encrypted."
+  }
+}
+```
+
+### Choosing Between Preconditions and PostConditions
+
+You can often implement a validation check as either a postcondition of the resource producing the data or as a precondition of a resource or output value using the data. To decide which is most appropriate, consider whether the check is representing either an assumption or a guarantee.
+
+| Type | Definition | Expression to Use |
+| -----|------------|--------------|
+| Assumption | A condition that must be true in order for the configuration of a particular resource to be usable. For example, an `aws_instance` configuration can have the assumption that the given AMI will always be configured for the `x86_64` CPU architecture. | Preconditions, so that future maintainers can find them close to the other expressions that rely on that condition. This lets them understand more about what that resource is intended to allow.|
+| Guarantee | A characteristic or behavior of an object that the rest of the configuration should be able to rely on. For example, an `aws_instance` configuration can have the guarantee that an EC2 instance will be running in a network that assigns it a private DNS record. | Postconditions, so that future maintainers can find them close to the resource configuration that is responsible for implementing those guarantees. This lets them more easily determine which behaviors they should preserve when changing the configuration. |
```

Comment:
Presenting this as a table does seem helpful for readability when I view it on my computer, but I worry this might be a bit awkward for those reading the page on a small-screen like a mobile device.

Were you able to test it there and verify it doesn't crunch up the columns too much, or force the user to scroll horizontally to read across the lines?

Reply:
as usual, an excellent callout, Martin. Let me do some testing/thinking on this. I agree this could be weird on a phone. I just feel like as headings the content is SUPER visually overwhelming...I'm going to think about this and come back with a better option

Reply:
table is gone - alas, you were right about it not looking great on phones. Thank you for calling that out - in my overzealousness to make the page more readable, I almost made it WORSE for mobile users.